x-pack/plugins/security_solution/public/onboarding/components/onboarding_body/hooks/use_completed_cards.ts

```diff
@@ -0,0 +1,139 @@
+      setCompleteCardIds((currentCompleteCards) => {
+        const isCurrentlyComplete = currentCompleteCards.includes(cardId);
+        if (completed && !isCurrentlyComplete) {
+          const newCompleteCardIds = [...currentCompleteCards, cardId];
+          reportCardComplete(cardId, options);
+          setStoredCompleteCardIds(newCompleteCardIds); // Keep the stored state in sync with the local state
+          return newCompleteCardIds;
+        } else if (!completed && isCurrentlyComplete) {
+          const newCompleteCardIds = currentCompleteCards.filter((id) => id !== cardId);
+          setStoredCompleteCardIds(newCompleteCardIds); // Keep the stored state in sync with the local state
+          return newCompleteCardIds;
+        }
+        return currentCompleteCards; // No change
+      });
+    },
+    [reportCardComplete, setStoredCompleteCardIds] // static dependencies, this function needs to be stable
+  );
+
+  const getCardCheckCompleteResult = useCallback<GetCardCheckCompleteResult>(
+    (cardId) => cardCheckCompleteResult[cardId],
+    [cardCheckCompleteResult]
+  );
+
+  // Internal: sets the checkCompleteResult for a specific card
+  const setCardCheckCompleteResult = useCallback(
+    (cardId: OnboardingCardId, options: CheckCompleteResult) => {
+      setCardsCompleteResult((currentCardCheckCompleteResult = {}) => ({
+        ...currentCardCheckCompleteResult,
+        [cardId]: options,
+      }));
+    },
+    []
+  );
+
+  // Internal: stores all cards that have a checkComplete function in a flat array
+  const cardsWithAutoCheck = useMemo(
+    () =>
+      cardsGroupConfig.reduce<OnboardingCardConfig[]>((acc, group) => {
+        acc.push(...group.cards.filter((card) => card.checkComplete));
+        return acc;
+      }, []),
+    [cardsGroupConfig]
+  );
+
+  // Internal: sets the result of a checkComplete function
+  const processCardCheckCompleteResult = useCallback(
+    (cardId: OnboardingCardId, checkCompleteResult: CheckCompleteResponse) => {
+      if (typeof checkCompleteResult === 'boolean') {
+        setCardComplete(cardId, checkCompleteResult, { auto: true });
+      } else {
+        const { isComplete, ...result } = checkCompleteResult;
+        setCardComplete(cardId, isComplete, { auto: true });
+        setCardCheckCompleteResult(cardId, result);
+      }
+    },
+    [setCardComplete, setCardCheckCompleteResult]
+  );
+
+  const checkCardComplete = useCallback(
+    (cardId: OnboardingCardId) => {
+      const cardConfig = cardsWithAutoCheck.find(({ id }) => id === cardId);
+
+      if (cardConfig) {
+        cardConfig.checkComplete?.(services).then((checkCompleteResult) => {
+          processCardCheckCompleteResult(cardId, checkCompleteResult);
+        });
+      }
+    },
+    [cardsWithAutoCheck, processCardCheckCompleteResult, services]
+  );
+
+  // Initial auto-check for all cards, it should run only once, after cardsGroupConfig is properly populated
```

Comment:
just double checking it says it should run once but is dependends on continuous identity checks?

Reply:
This effect should run only once after the configs have been populated. It triggers twice, one with an empty array, and another with the populated items. It won't trigger again.
 
Nevertheless, I changed the implementation a bit to make it more obvious and resilient to changes.